### PR TITLE
feat(lsp): npm specifier completions

### LINKED ISSUE #15811: lsp: npm registry completions on module specifiers
The LSP should support npm registry completions on module specifiers.

### PR DESCRIPTION
**Pull request by nayeemrmn**
Closes #15811.

Snipping tool screen recorder isn't working so here's some screenshots:
image
image

### CODE REVIEW

**Comment by nayeemrmn** [comment]
What kind of test would we allow for this?

**Comment by ry** [comment]
Nice work. I think there are some existing completion tests in the LSP code that could be copied for this?

**Comment by nayeemrmn** [comment]
This one depends on the npm registry site... I think I'll swap out the endpoint URLs in a `cfg(test)` to a localhost with fake responses.

**Comment by bartlomieju** [comment]
> This one depends on the npm registry site... I think I'll swap out the endpoint URLs in a `cfg(test)` to a localhost with fake responses.

That's a great idea.

**Comment by nayeemrmn** [comment]
Okay we already have a `CliNpmRegistryApi` abstraction whose origin can be overrided with an env var which npm tests use.

**Comment by nayeemrmn** [comment]
I wanted to reuse the loading stuff in `CliNpmRegistryApi` but when I tried to move the search API request over there was a lot of concurrency and cache machinery that looked overkill for the LSP. That means there are now two spots where package info is fetched.

File: cli/lsp/completions.rs [hunk 1]
```diff
@@ -452,6 +471,134 @@ fn get_relative_specifiers(
     .collect()
 }
 
+/// Get completions for `npm:` specifiers.
+async fn get_npm_completions(
+  specifier: &str,
+  range: &lsp::Range,
+  npm_registry_url: &Url,
+  file_fetcher: &FileFetcher,
```
Comment (dsherret) [inline cli/lsp/completions.rs:479]: Instead of providing a file fetcher and registry url, maybe we should create and provide an `NpmSearchApi` struct here instead that would have the file fetcher and `npm_registry_url` as being internal to it. Then that would allow us to easily create an `NpmSearchApi` trait (with `RealNpmSearchApi` struct) that we could then easily implement for testing purposes.

**Review by dsherret — COMMENTED** [review]
Thanks! This will be a big improvement.

File: cli/lsp/completions.rs [hunk 2]
```diff
@@ -452,6 +471,134 @@ fn get_relative_specifiers(
...
+    url
+      .path_segments_mut()
+      .ok()?
+      .pop_if_empty()
+      .extend(suffix.split('/'));
+    Some(url)
+  };
+
+  // First try to match `npm:some-package@<version-to-complete>`.
+  let v_index = bare_specifier.rfind('@');
```
[the remark this answers is not in the document]
Comment (nayeemrmn) [reply cli/lsp/completions.rs]: > Perhaps instead of this we should use `NpmPackageReqReference` to try to parse the npm specifier first?

Not sure if we can use that since it validates for a completed specifier. But I made the parsing better.

File: cli/lsp/completions.rs (hunk 1 above)
Comment (nayeemrmn) [reply cli/lsp/completions.rs]: I merged it into `CliNpmRegistryApi` after all.

File: cli/npm/registry.rs [hunk 3]
```diff
@@ -95,6 +97,14 @@ impl CliNpmRegistryApi {
     self.inner().get_cached_package_info(name)
   }
 
+  // TODO(nayeemrmn): Maybe declare in `NpmRegistryApi`.
```
Comment (dsherret) [inline cli/npm/registry.rs:100]:
```suggestion
```

The `NpmRegistryApi` trait in `deno_npm` shouldn't have this function because the trait should only have functionality that `deno_npm` needs and we don't want to require implementors to implement this when it's not used there.

File: cli/lsp/language_server.rs [hunk 4]
```diff
@@ -481,10 +481,11 @@ impl LanguageServer {
   }
 }
 
-fn create_npm_api_and_cache(
+pub fn create_npm_api_and_cache(
```
Comment (dsherret) [inline cli/lsp/language_server.rs:484]: This is a high level function used for construction and we shouldn't expose it to the rest of the application. We can instead find a way to break down the components to make them more usable in tests and revert the changes here.

File: cli/lsp/completions.rs [hunk 5]
```diff
@@ -682,6 +800,232 @@ mod tests {
     );
   }
 
+  #[tokio::test]
+  async fn test_get_npm_completions() {
```
Comment (dsherret) [inline cli/lsp/completions.rs:804]: The tests here are very high level. For example, it's creating an http server, temp dir, http client, progress bars, npm api, cache. We can slim this down a lot for unit testing purposes and then maybe add a single basic integration test in `tests/integration/lsp_tests.rs`. The registry response is also far away from these tests in test_util/src/lib.rs which makes it harder to follow.

We could improve this by making a seam for the npm api like so:

```
#[async_trait::async_trait]
pub trait NpmSearchApi {
  async fn search(
    &self,
    query: &str,
  ) -> Result<Arc<Vec<String>>, AnyError>;
}

async fn get_npm_completions(
  specifier: &str,
  range: &lsp::Range,
  npm_api: &impl NpmSearchApi,
) -> Option<Vec<lsp::CompletionItem>> {
```

Then in the tests we can inject responses for different queries:

```rs
struct FakeNpmSearchApi(HashMap<String, Result<Arc<Vec<String>>, String>>);

#[async_trait::async_trait]
impl NpmSearchApi for FakeNpmSearchApi {
  async fn search(
    &self,
    query: &str,
  ) -> Result<Arc<Vec<String>>, AnyError> {
    self.0.get(query) // etc... with actual code that works here
  }
}
```

Then we could implement this trait for whatever struct we want in the real CLI which could then be used for a basic integration test to ensure all the plumbing is hooked up correctly or anything else that should be tested at an integration level.

This would allow us to keep everything close together and wouldn't require tying the lsp setup code to these tests.

File: cli/npm/registry.rs [hunk 6]
```diff
@@ -397,4 +409,52 @@ impl CliNpmRegistryApiInner {
...
+    search_url
+      .path_segments_mut()
+      .map_err(|_| anyhow!("Custom npm registry URL cannot be a base."))?
+      .pop_if_empty()
+      .extend("-/v1/search".split('/'));
+    search_url
+      .query_pairs_mut()
+      .append_pair("text", &format!("{} boost-exact:false", query));
+
+    let bytes = self.http_client.download(search_url).await?;
```
Comment (dsherret) [inline cli/npm/registry.rs:445]: In the previous commit this was using `ModuleRegistry`'s file fetcher. Is it ok for this to be using the http client directly here (I think it's probably ok?)? If not, maybe we should just move this functionality out of `CliNpmRegistryApi` and make a more targeted `NpmSearchApi` for the lsp in order to not accidentally mix the two together and reduce the dependencies necessary to create a `CliNpmRegistryApi`.

File: cli/lsp/registries.rs [hunk 7]
```diff
@@ -415,7 +415,7 @@ enum VariableItems {
 #[derive(Debug, Clone)]
 pub struct ModuleRegistry {
   origins: HashMap<String, Vec<RegistryConfiguration>>,
-  file_fetcher: FileFetcher,
+  pub file_fetcher: FileFetcher,
```
Comment (dsherret) [inline cli/lsp/registries.rs:418]:
```suggestion
  file_fetcher: FileFetcher,
```

**Review by dsherret — COMMENTED** [review]
Almost there. Given the changes made that expose the high level construction code to the unit tests, I think we should make a seam for the api provided to `get_npm_completions` so we don't need to expose `create_npm_api_and_cache`.

File: cli/lsp/completions.rs (hunk 5 above)
Comment (nayeemrmn) [reply cli/lsp/completions.rs:821]: Ended up with:
```rust
async fn get_npm_completions(
  specifier: &str,
  range: &lsp::Range,
  npm_registry_api: &impl NpmRegistryApi,
  npm_search_api: &impl NpmSearchApi,
) -> Option<Vec<lsp::CompletionItem>> {}
```
since search and package metadata are separate. Also it's still using `CliNpmRegistryApi` as the implementer for `NpmSearchApi` for convenience

File: cli/lsp/completions.rs [hunk 8]
```diff
@@ -452,6 +463,114 @@ fn get_relative_specifiers(
...
+      .filter(|idx| !bare_specifier[..*idx].is_empty())
+      .filter(|idx| !bare_specifier[..*idx].contains('/'))
+  };
+
+  // First try to match `npm:some-package@<version-to-complete>`.
+  if let Some(v_index) = v_index {
+    let package_name = &bare_specifier[..v_index];
+    let v_prefix = &bare_specifier[(v_index + 1)..];
+    let versions = &npm_registry_api
+      .package_info(package_name)
```
Comment (dsherret) [inline cli/lsp/completions.rs:499]: The implementation of this currently uses the cached package information in the LSP. It would be good to make the `NpmSearchApi` trait have a `async fn package_versions(&self) -> Option<Vec<Version>>` method so that this is not coupled to `deno_npm`'s trait and then we could implement a more targeted search api that also uses the module registry's file fetcher.

**Review by dsherret — CHANGES_REQUESTED** [review]
Sorry for my delay on this one! I thought it was good and we were just waiting for the merge window to open, but then I just looked at it a bit more and I think it will only use the cached package information. Additionally, we could decouple `get_npm_completions` from `deno_npm`'s `NpmRegistryApi` trait.

**Review by dsherret — APPROVED** [review]
Looks great. Nice work and thanks @nayeemrmn!
